roachtest/pg_regress: update and improve the runner

yuzefovich opened the pull request:
This PR improves the pg_regress setup by splitting the large `regression.diffs` file into a separate diff file for each sub-test which should make reviewing of the changes going forward easier. Additionally it adds automatic handling of `_version_` markers. It unskips `oidjoins` file.

It also accepts diffs that have been accumulated over the last several months. Several issues have been filed to track separately.

Please see each commit for more details.

Fixes: #137482

Release note: None

yuzefovich committed:
roachtest/pg_regress: improve the runner

This commit adjusts the pg_regress test so that `regression.diffs` file
contains generic "_version_" markers removing some manual work when
a new major version is released. It is achieved by initially copying the
result into a temp file, then doing the same post-processing as we
already did to replace the version with the generic marker, and then
writing the updated diff into `regression.diffs`.

It also skips `prepared_xacts` file that is hanging due to a known
issue.

Release note: None

yuzefovich committed:
roachtest/pg_regress: update the regression diff

I didn't review the changes here very closely and simply accepted all of
them. Several issues have been filed to track identified problems (two
about new assertion failures, and two about a regression around roles).
Notable changes since the last update that was several months ago:
- changes to interval math
- some error messages changed
- work is happening on PREPARE TRANSACTION
- some new ALTER COLUMN TYPE support
- DO blocks are now supported
- holdable cursors
- jsonpath support.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142774)
<!-- Reviewable:end -->

yuzefovich committed:
roachtest/pg_regress: split up regression.diffs into separate files

This commit replaces a single run over all test files in `pg_regress`
with multiple runs over each test file separately. This results in
breaking up huge `regression.diffs` file into lots of smaller ones,
which should make reviewing the changes easier going forward.

The downside is that now multiple files might need to be copied from
the artifacts whenever the diff changes in expected manner, but I
think the ease of reviewing far outweighs this.

Some other smaller downsides are:
- the test takes longer (about 27 minutes instead of 10 minutes before)
- we no longer get a single `regression.out` file which indicates which
files we fully pass (we don't yet really pay attention to this anyway).

Release note: None

yuzefovich committed:
roachtest/pg_regress: enable oidjoins file

This was disabled with a comment to enable once we support DO blocks
which we now do.

Release note: None

yuzefovich commented:
Hold off on reviewing for a sec - I'll make a change so that each sub test resulted in a separate diff file.

yuzefovich commented:
Ok, should be RFAL.

Note that reviewing the newly-accepted diffs in the second commit is a pain, so I'd suggest not spending lots of time on that - we do have large diffs anyway, and I filed a couple of issues for new internal errors. In other commits, only `pg_regress.go` and `lint_test.go` should have meaningful changes.

rafiss commented on pkg/cmd/roachtest/tests/pg_regress.go:
```diff
@@ -472,6 +497,8 @@ func registerPGRegress(r registry.Registry) {
 		Suites:            registry.Suites(registry.Weekly),
 		Leases:            registry.MetamorphicLeases,
+		// TODO(#142777): matview file results in an invalid descriptor.
+		SkipPostValidations: registry.PostValidationInvalidDescriptors,
```
the bug has been fixed: https://github.com/cockroachdb/cockroach/pull/143287

yuzefovich replied:
Thanks!

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 3 of 191 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142774#-:-OM7bKrA0VzXzRUeJc7P:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball, @michae2, and @yuzefovich)



<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 191 of 191 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142774#-:-OM88W-5BPMVY8viyW24:bnfp4nl)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @michae2 and @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich committed:
roachtest/pg_regress: accept diffs from last two weeks

Release note: None

yuzefovich commented:
TFTRs!

bors r+
